src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -336,7 +345,17 @@ private bool IsRecovering(ServerErrorCode code, string message)
 
             return false;
         }
-        
+
+        private bool ShouldClearConnectionPoolForChannelException(Exception ex, SemanticVersion serverVersion)
+        {
+            if (ex is MongoNotPrimaryException mongoNotPrimaryException && mongoNotPrimaryException.Code == (int)ServerErrorCode.NotMaster)
+            {
+                return !Feature.KeepConnectionPoolWhenNotMasterConnectionException.IsSupported(serverVersion);
```

Comment:
I think this checking can be replaced on a validation for a `MaxWireProtocol` version  since this:

    Proposal: On "not master", look up the server's maxWireVersion from the TopologyDescription. If maxWireVersion indicates MongoDB 4.2+, don't clear the pool. Leave the rest of this spec section unchanged.

https://jira.mongodb.org/browse/SPEC-1276

Reply:
Also, maybe we can add more cases which gives the method result as `true` explicitly?